src/libraries/System.Net.Primitives/tests/FunctionalTests/IPEndPointParsing.cs

```diff
@@ -301,5 +303,337 @@ private void InvalidPortHelper(string addressAndPort)
+        [InlineData("[0:0:111:234:5:6:789A:0]:443", "[::111:234:5:6:789a:0]:443")]
+        [InlineData("E:E:E:E:E:E:0:1", "[e:e:e:e:e:e:0:1]:0")]
+        [InlineData("[E:E:E:E:E:E:0:1]:443", "[e:e:e:e:e:e:0:1]:443")]
+        public static void TryFormatByteSpan(string input, string expected)
+        {
+            Span<byte> destination = stackalloc byte[expected.Length];
+
+            Assert.True(IPEndPoint.Parse(input).TryFormat(destination, out int charsWritten));
+            Assert.Equal(expected.Length, charsWritten);
+            Assert.True(MemoryExtensions.SequenceEqual(destination, Encoding.UTF8.GetBytes(expected)));
+        }
+
+        [Theory]
+        [InlineData("[::1]", "[::1]:0")]
+        [InlineData("[::1]:5040", "[::1]:5040")]
+        [InlineData("10.12.13.14", "10.12.13.14:0")]
+        [InlineData("10.12.13.14:5040", "10.12.13.14:5040")]
+        [InlineData("0:0:111:234:5:6:789A:0", "[::111:234:5:6:789a:0]:0")]
+        [InlineData("[0:0:111:234:5:6:789A:0]:443", "[::111:234:5:6:789a:0]:443")]
+        [InlineData("E:E:E:E:E:E:0:1", "[e:e:e:e:e:e:0:1]:0")]
+        [InlineData("[E:E:E:E:E:E:0:1]:443", "[e:e:e:e:e:e:0:1]:443")]
+        public static void ToFormatIFormattable(string input, string expectedFormat)
+        {
+            IPEndPoint ep = IPEndPoint.Parse(input);
+            string result = string.Format("display {0:G}", ep);
+
+            Assert.Equal($"display {expectedFormat}", result);
+        }
+
+        private static class ParsableHelper<TSelf> where TSelf : IParsable<TSelf>
+        {
+            public static TSelf Parse(string s, IFormatProvider provider) => TSelf.Parse(s, provider);
+
+            public static bool TryParse(string? s, IFormatProvider provider, out TSelf result) => TSelf.TryParse(s, provider, out result);
+        }
+
+        [Theory]
+        [InlineData("[::1]", "::1", 0)]
+        [InlineData("[::1]:5040", "::1", 5040)]
+        [InlineData("10.12.13.14", "10.12.13.14", 0)]
+        [InlineData("10.12.13.14:5040", "10.12.13.14", 5040)]
+        [InlineData("0:0:111:234:5:6:789A:0", "::111:234:5:6:789a:0", 0)]
+        [InlineData("[0:0:111:234:5:6:789A:0]:443", "::111:234:5:6:789a:0", 443)]
+        [InlineData("E:E:E:E:E:E:0:1", "e:e:e:e:e:e:0:1", 0)]
+        [InlineData("[E:E:E:E:E:E:0:1]:443", "e:e:e:e:e:e:0:1", 443)]
+        public static void ParseIParsable(string input, string address, int port)
+        {
+            IPEndPoint result = ParsableHelper<IPEndPoint>.Parse(input, CultureInfo.InvariantCulture);
+            Assert.Equal(port, result.Port);
+            Assert.Equal(address, result.Address.ToString());
+        }
+
+        [Fact]
+        public static void ParseNullIParsable_ThrowsArgumentNullException()
+        {
+            Assert.Throws<ArgumentNullException>(() => ParsableHelper<IPEndPoint>.Parse(null, CultureInfo.InvariantCulture));
+        }
+
+        [Theory]
+        [InlineData("10.12.13.-14")]
```

Comment:
If some of these input data are the same for multiple tests, they could be extracted and shared via `MemberData`.